topaz/extract.py

```diff
@@ -5,69 +5,188 @@
+def nms_iterator(scores:Iterable[np.ndarray], radius:int, threshold:float, pool:multiprocessing.Pool=None, dims:int=3, chunk_size:int=128) -> Iterator[tuple[str, np.ndarray, np.ndarray]]:
+    """
+    Iterator for performing non-maximum suppression on a batch of scores.
+
+    Args:
+        scores (Iterable[np.ndarray]): Iterable of score arrays.
+        radius (int): Radius for suppression.
+        threshold (float): Threshold for suppression.
+        pool (multiprocessing.Pool, optional): Multiprocessing pool for parallel processing.
+        dims (int): Number of dimensions (2 or 3).
+        chunk_size (int): Size of chunks for processing large 3D volumes.
+
+    Yields:
+        tuple: Tuple containing name, suppressed score, and coordinates.
+    """
     process = NonMaximumSuppression(radius, threshold, dims=dims)
     if pool is not None:
-        for name,score,coords in pool.imap_unordered(process, scores):
-            yield name,score,coords
+        for name, score in scores:
+            print(f"Debug: Initial score shape: {score.shape}")
+            original_shape = score.shape
+            if score.ndim != 3:
+                print(f"Warning: Expected 3D score array, but got shape {score.shape}. Attempting to reshape...")
+                if score.ndim == 1:
+                    side = int(np.cbrt(score.size))
+                    score = score.reshape(side, side, side)
+                elif score.ndim == 2:
+                    score = score.reshape(1, *score.shape)
+                print(f"Reshaped score array to {score.shape}")
+
+            coords_list = []
+            scores_list = []
+            for z in range(0, score.shape[0], chunk_size):
+                for y in range(0, score.shape[1], chunk_size):
+                    for x in range(0, score.shape[2], chunk_size):
+                        z_end = min(z + chunk_size, score.shape[0])
+                        y_end = min(y + chunk_size, score.shape[1])
+                        x_end = min(x + chunk_size, score.shape[2])
+                        
+                        chunk = score[z:z_end, y:y_end, x:x_end]
+                        _, chunk_score, chunk_coords = process((name, chunk))
+                        
+                        # Adjust coordinates to reflect position in full tomogram
+                        chunk_coords[:, 0] += x  # x coordinate
+                        chunk_coords[:, 1] += y  # y coordinate
+                        chunk_coords[:, 2] += z  # z coordinate
+                        
+                        coords_list.append(chunk_coords)
+                        scores_list.append(chunk_score)
+            
+            coords = np.concatenate(coords_list, axis=0) if coords_list else np.array([])
+            score = np.concatenate(scores_list, axis=0) if scores_list else np.array([])
+            
+            print(f"Debug: Final score shape: {score.shape}, coords shape: {coords.shape}")
```

Comment:
Score patching code is here, a few TODOs:

- Support 2d input
- Make patching optional
- Add padding to each patch to remove edge artifacts
- Add patching below to else when not using a process pool
- Patching _could_ be pushed down to the score layer so that the scoring network is also applied patch-by-patch and pipelined directly into the patched NMS algorithm, but might not be worth the trouble